### DISCUSSION TITLE
basis array allocation with much larger size than actual basis.Ns

### DISCUSSION THREAD

**Discussion by tpyang** [QuSpin/QuSpin#545] [Q&A] [answered]
I am trying to create a spinful_fermion_basis_general as following:
basis_2d = spinful_fermion_basis_general(27,Nf=(21,3),double_occupancy=False)

While the basis size should be:
import scipy.special as ss
int(ss.comb(27,24))*int(ss.comb(24,3)) = 5920200

I add one line: print('creating basis dimension', Ns, self._Ns, Ns_block_est, self._Ns_block_est)
before line 746 in site-packages/quspin/basis/basis_general/base_general.py:
basis = basis_zeros(Ns,dtype=self._basis_dtype)

Here is what I got in the output:
creating basis dimension 3463317000 3463317000 None 3463317000

Could I ask what's the reason for much larger Ns (should be 5920200, if you check basis_2d.Ns ) in the creating process? And can this be avoided? 

Thank you very much!

**Comment by weinbe58** [comment, marked as answer]
it is estimating the size of the basis with double occupancy. 

If you know a good way to get the size of the non-symmetry reduced basis we can add it to the current implementation. Otherwise you have to manually put in an estimate with the argument: Ns_block_est=...

**Reply by tpyang** [reply]
Thank you so much for the quick answer! 

In the spin full fermion basis, the estimation of the size of non-symmetry reduced basis without double occupancy is straightforward. We get the size as:
Ns_block_est= int(ss.comb(Nsite, Nup+Ndn))*int(ss.comb(Nup+Ndn,Ndn))

In my limited test cases, I find the final basis is independent of the Ns_block_est.  Just to confirm, setting Ns_block_est would not affect the final construction of basis?

**Reply by tpyang** [reply]
Sorry for the long follow up! But your answer already solved my concern.
